File: medium/graph/noOfIslands.cpp
#include<iostream>
#include<vector>
#include<cmath> 
#include<algorithm>
#include<unordered_set>
#include<set>
#include<string>
#include<map>
#include<unordered_map>
#include<cstring>
#include<stack>
#include<queue>
#include<climits>

using namespace std;

#define ll long long

const int INF = 1e9+7;
const int mod = 1e9+7;
const int MAX = 1e5+5;

bool vis[1234][1234];

vector<int> rowNbr{-1,0,0,1};
vector<int> colNbr{0,-1,1,0};

bool isSafe(int x, int y, vector<vector<char> >grid){
	if(x < grid.size() && x >= 0 && y < grid[0].size() && y >= 0)
		return true;
	return false;
}

void dfs(int row, int col, vector<vector<char>> grid){
	vis[row][col] = true;
	for(int i = 0; i < 4; ++i){
		int x = row + rowNbr[i];
		int y = col + colNbr[i];
		if(isSafe(x, y, grid) && !vis[x][y] && grid[x][y] == '1'){
			dfs(x, y, grid);
		} 
	}
}

int numIslands(vector<vector<char>> &grid){
	int islands = 0;
	for(int i = 0; i < grid.size(); ++i){
		for(int j = 0; j < grid[0].size(); ++j){
			if(!vis[i][j] && grid[i][j] == '1'){
				dfs(i, j, grid);
				islands++;
			}		
		}
	}
	return islands;
}

int main(){
	//freopen("tree.in","r",stdin);
    //freopen("tree.out","w",stdout);
    //freopen("input.txt", "r", stdin);
    //freopen("out.txt", "w", stdout);
    ios_base::sync_with_stdio(0);
    cin.tie(0);

 	vector<vector<char> > grid{ {'1','1','1','1','0'}, 
 								{'1','1','0','1','0'},
 								{'1','1','0','0','0'}, 
 								{'0','0','0','0','0'} };
 	cout << numIslands(grid) << endl;
}

File: medium/graph/noOfIslands.py
vis = [[0 for j in range(1234)] for i in range(1234)]

rowNbr = [-1, 0, 0, 1]
colNbr = [0, -1, 1, 0]

def isSafe(x, y, grid):
	return True if x >= 0 and x < len(grid) and y >= 0 and y < len(grid[0]) else False

def dfs(row, col, grid, vis):
	vis[row][col] = 1
	for i in range(4):
		x = row+rowNbr[i]
		y = col+colNbr[i]
		if isSafe(x, y, grid) and (not vis[x][y]) and grid[x][y] == '1':
			dfs(x, y, grid, vis)

def noOfIslands(grid):
	islands = 0
	m = len(grid)
	n = len(grid[0])
	# global vis; rowNbr; colNbr

	for i in range(m):
		for j in range(n):
			if not vis[i][j] and grid[i][j] == '1':
				dfs(i, j, grid, vis)
				islands += 1

	return islands		

if __name__ == '__main__':
	grid = [['1','1','0','0','0'], 
			['0','0','0','0','0'],
			['0','0','1','0','0'], 
			['0','0','0','1','1'] ]
	print(noOfIslands(grid))		